File: apps/mobile/test/product/money/live_trade_history_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:trimmy/account/account_controller.dart';
import 'package:trimmy/practice_sync/http_transport.dart';
import 'package:trimmy/product/design/product_theme.dart';
import 'package:trimmy/product/market/live_trading.dart';
import 'package:trimmy/product/money/live_trade_history.dart';

import '../../support/account_data_fixtures.dart' as fixtures;

final _origin = Uri.parse('https://trimmy.example');
String _id(int value) =>
    '11111111-1111-4111-8111-${value.toString().padLeft(12, '0')}';
Map<String, Object?> _trade(
  int id, {
  String status = 'confirmed',
  bool buy = true,
}) => {
  'id': _id(id),
  'wallet': fixtures.wallet,
  'status': status,
  'signature': '2' * 88,
  'createdAt': '2026-09-25T13:00:00.123456Z',
  'updatedAt': '2026-09-25T13:00:10.123456Z',
  'asset': {
    'assetId': 'nvidia',
    'mint': fixtures.nvidiaMint,
    'symbol': 'NVDAx',
    'name': 'NVIDIA',
    'decimals': 8,
  },
  'terms': <String, Object?>{
    'side': buy ? 'buy' : 'sell',
    'inputMint': buy ? liveUsdcMint : fixtures.nvidiaMint,
    'outputMint': buy ? fixtures.nvidiaMint : liveUsdcMint,
    'inputAmountRaw': buy ? '5000000' : '12345678',
    'quotedOutputAmountRaw': buy ? '12345678' : '5000000',
    'minimumOutputAmountRaw': buy ? '12000000' : '4900000',
  },
  'amountUnits': 'raw_token_units',
  'amountsStatus': 'reviewed_quote',
};
Map<String, Object?> _page(List<Map<String, Object?>> rows, {String? cursor}) =>
    {
      'schemaVersion': 1,
      'network': 'solana:mainnet-beta',
      'orders': rows,
      'nextCursor': cursor,
    };
http.Response _reply(Object? value, {int status = 200}) => http.Response(
  jsonEncode(value),
  status,
  headers: {'content-type': 'application/json'},
);

class _Account extends ChangeNotifier implements AccountController {
  @override
  String? accountId = fixtures.account;
  @override
  int navigationEpoch = 0;
  @override
  AccountPhase phase = AccountPhase.active;
  Completer<PracticeAccessToken>? tokenCompleter;
  @override
  Future<PracticeAccessToken> freshAccessToken() async =>
      tokenCompleter?.future ??
      PracticeAccessToken(accountId: accountId!, token: 'test-token');
  void changeAccount() {
    accountId = 'another-account';
    navigationEpoch++;
    notifyListeners();
  }

  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

Future<void> _pump(WidgetTester tester) async {
  await tester.runAsync(() async {
    await Future<void>.delayed(Duration.zero);
  });
  await tester.pump();
  await tester.runAsync(() async {
    await Future<void>.delayed(Duration.zero);
  });
  await tester.pump(const Duration(milliseconds: 1800));
}

Future<void> _mount(
  WidgetTester tester,
  _Account account,
  http.Client client, {
  Future<bool> Function(Uri)? explorer,
}) async {
  await tester.pumpWidget(
    MaterialApp(
      theme: productTheme(),
      home: LiveTradeHistoryScreen(
        account: account,
        origin: _origin,
        onBack: () {},
        httpClient: client,
        openExplorer: explorer,
      ),
    ),
  );
  await _pump(tester);
}

Future<void> _close(WidgetTester tester, _Account account) async {
  await tester.pumpWidget(const SizedBox());
  account.dispose();
}

void main() {
  test(
    'history preserves exact quantities and labels quotes instead of fills',
    () {
      final page = LiveTradeHistoryPage.fromJson(
        _page([_trade(1), _trade(2, buy: false)]),
      );
      expect(page.orders.first.inputLabel, '5 USDC');
      expect(page.orders.first.quotedOutputLabel, '0.12345678 NVDAx raw units');
      expect(page.orders.last.inputLabel, '0.12345678 NVDAx raw units');
      expect(page.orders.last.minimumOutputLabel, '4.9 USDC');
      expect(page.orders.first.createdAt.microsecond, 456);
      expect(page.orders.first.explorer.host, 'solscan.io');
    },
  );

  test(
    'malformed amounts, sides, networks and duplicate records fail closed',
    () {
      final mutations = <void Function(Map<String, Object?>)>[
        (row) => row['status'] = 'reviewed',
        (row) => row['signature'] = null,
        (row) => row['signature'] = 'https://other.test',
        (row) => row['amountsStatus'] = 'final_fill',
        (row) => (row['terms'] as Map)['inputMint'] = fixtures.wallet,
        (row) => (row['terms'] as Map)['inputAmountRaw'] = 5000000,
        (row) =>
            (row['terms'] as Map)['inputAmountRaw'] = '18446744073709551616',
        (row) => (row['terms'] as Map)['minimumOutputAmountRaw'] = '999999999',
        (row) => (row['asset'] as Map)['decimals'] = 19,
        (row) => row['createdAt'] = '2026-09-25T13:00:00',
      ];
      for (final mutate in mutations) {
        final row = _trade(1);
        mutate(row);
        expect(
          () => LiveTradeHistoryPage.fromJson(_page([row])),
          throwsFormatException,
        );
      }
      expect(
        () => LiveTradeHistoryPage.fromJson({
          ..._page([]),
          'network': 'solana:devnet',
        }),
        throwsFormatException,
      );
      expect(
        () => LiveTradeHistoryPage.fromJson(_page([_trade(1), _trade(1)])),
        throwsFormatException,
      );
      expect(
        () => LiveTradeHistoryPage.fromJson(_page([], cursor: 'more')),
        throwsFormatException,
      );
    },
  );

  test(
    'history is authenticated GET with escaped cursor and never follows redirects',
    () async {
      final account = _Account();
      final client = LiveTradeHistoryClient(
        account: account,
        origin: _origin,
        client: MockClient((request) async {
          expect(request.method, 'GET');
          expect(request.url.path, '/v1/trading/history');
          expect(request.url.queryParameters, {
            'limit': '20',
            'cursor': 'opaque+/=cursor',
          });
          expect(request.headers['authorization'], 'Bearer test-token');
          expect(request.followRedirects, isFalse);
          return _reply(_page([_trade(1)]));
        }),
      );
      expect((await client.read(cursor: 'opaque+/=cursor')).orders.length, 1);
      client.close();
      account.dispose();
    },
  );

  test(
    'account switch during token acquisition prevents authenticated dispatch',
    () async {
      final account = _Account()
        ..tokenCompleter = Completer<PracticeAccessToken>();
      var dispatched = 0;
      final client = LiveTradeHistoryClient(
        account: account,
        origin: _origin,
        client: MockClient((_) async {
          dispatched++;
          return _reply(_page([]));
        }),
      );
      final result = client.read();
      final failure = expectLater(
        result,
        throwsA(
          isA<LiveTradeHistoryFailure>().having(
            (e) => e.code,
            'code',
            'ACCOUNT_REQUIRED',
          ),
        ),
      );
      account.changeAccount();
      account.tokenCompleter!.complete(
        PracticeAccessToken(accountId: fixtures.account, token: 'old-token'),
      );
      await failure;
      expect(dispatched, 0);
      client.close();
      account.dispose();
    },
  );

  test(
    'old account response is discarded after an in-flight account switch',
    () async {
      final account = _Account();
      final reply = Completer<http.Response>();
      final sent = Completer<void>();
      final client = LiveTradeHistoryClient(
        account: account,
        origin: _origin,
        client: MockClient((_) {
          sent.complete();
          return reply.future;
        }),
      );
      final failure = expectLater(
        client.read(),
        throwsA(
          isA<LiveTradeHistoryFailure>().having(
            (e) => e.code,
            'code',
            'ACCOUNT_REQUIRED',
          ),
        ),
      );
      await sent.future;
      account.changeAccount();
      reply.complete(_reply(_page([_trade(1)])));
      await failure;
      client.close();
      account.dispose();
    },
  );

  test('oversized responses and nonadvancing cursors are rejected', () async {
    final account = _Account();
    var oversized = true;
    final client = LiveTradeHistoryClient(
      account: account,
      origin: _origin,
      client: MockClient(
        (_) async => oversized
            ? http.Response('x' * 131073, 200)
            : _reply(_page([_trade(1)], cursor: 'same')),
      ),
    );
    await expectLater(client.read(), throwsA(isA<LiveTradeHistoryFailure>()));
    oversized = false;
    await expectLater(client.read(cursor: 'same'), throwsFormatException);
    client.close();
    account.dispose();
  });

  testWidgets(
    'history displays status and exact quoted details with a safe explorer link',
    (tester) async {
      final account = _Account();
      Uri? opened;
      await _mount(
        tester,
        account,
        MockClient(
          (_) async => _reply(_page([_trade(1), _trade(2, status: 'failed')])),
        ),
        explorer: (uri) async {
          opened = uri;
          return true;
        },
      );
      expect(find.text('Confirmed'), findsOneWidget);
      expect(find.text('Not completed'), findsOneWidget);
      expect(find.textContaining('paper'), findsNothing);
      await tester.tap(find.byKey(ValueKey('trade-${_id(1)}')));
      await tester.pump();
      expect(find.text('Quoted output'), findsOneWidget);
      expect(find.text('0.12345678 NVDAx raw units'), findsOneWidget);
      await tester.tap(find.text('View transaction'));
      await _pump(tester);
      expect(opened, Uri.https('solscan.io', '/tx/${'2' * 88}'));
      await _close(tester, account);
    },
  );

  testWidgets(
    'pagination failure retains rows and retries without duplicates',
    (tester) async {
      final account = _Account();
      var moreCalls = 0;
      await _mount(
        tester,
        account,
        MockClient((request) async {
          if (!request.url.queryParameters.containsKey('cursor')) {
            return _reply(_page([_trade(1)], cursor: 'older'));
          }
          moreCalls++;
          return moreCalls == 1
              ? _reply({'code': 'HISTORY_UNAVAILABLE'}, status: 503)
              : _reply(_page([_trade(1), _trade(2, buy: false)]));
        }),
      );
      await tester.tap(find.text('More trades'));
      await _pump(tester);
      expect(find.byKey(ValueKey('trade-${_id(1)}')), findsOneWidget);
      expect(
        find.text('Couldn’t load your trades. Try again.'),
        findsOneWidget,
      );
      await tester.tap(find.text('More trades'));
      await _pump(tester);
      expect(find.byKey(ValueKey('trade-${_id(1)}')), findsOneWidget);
      expect(find.byKey(ValueKey('trade-${_id(2)}')), findsOneWidget);
      expect(find.text('More trades'), findsNothing);
      await _close(tester, account);
    },
  );

  testWidgets(
    'retry recovers the unavailable state and empty history stays honest',
    (tester) async {
      final account = _Account();
      var calls = 0;
      await _mount(
        tester,
        account,
        MockClient((_) async {
          calls++;
          return calls == 1 ? _reply({}, status: 503) : _reply(_page([]));
        }),
      );
      expect(find.text('Try again'), findsOneWidget);
      await tester.tap(find.text('Try again'));
      await _pump(tester);
      expect(find.text('Your first trade starts here'), findsOneWidget);
      await _close(tester, account);
    },
  );

  testWidgets(
    'pending history reconciles through GET and becomes confirmed automatically',
    (tester) async {
      final account = _Account();
      var resolved = false, reconciles = 0;
      await _mount(
        tester,
        account,
        MockClient((request) async {
          expect(request.method, 'GET');
          if (request.url.path.startsWith('/v1/trading/order/')) {
            reconciles++;
            resolved = true;
            return _reply({
              'order': {'id': _id(1), 'status': 'confirmed'},
            });
          }
          return _reply(
            _page([_trade(1, status: resolved ? 'confirmed' : 'pending')]),
          );
        }),
      );
      expect(find.text('Confirming'), findsOneWidget);
      await tester.pump(const Duration(seconds: 10));
      await _pump(tester);
      expect(reconciles, 1);
      expect(find.text('Confirmed'), findsOneWidget);
      await _close(tester, account);
    },
  );

  testWidgets('account changes immediately remove the previous user history', (
    tester,
  ) async {
    final account = _Account();
    await _mount(
      tester,
      account,
      MockClient((_) async => _reply(_page([_trade(1)]))),
    );
    expect(find.text('Buy NVDAx'), findsOneWidget);
    account.changeAccount();
    await tester.pump();
    expect(find.text('Buy NVDAx'), findsNothing);
    expect(find.text('Sign in again to see your trades.'), findsOneWidget);
    await _close(tester, account);
  });

  testWidgets(
    'pagination during pending reconciliation cannot strand Loading',
    (tester) async {
      final account = _Account();
      final reconciliation = Completer<http.Response>();
      final older = Completer<http.Response>();
      var reconciling = false, paging = false;
      await _mount(
        tester,
        account,
        MockClient((request) async {
          if (request.url.path.startsWith('/v1/trading/order/')) {
            reconciling = true;
            return reconciliation.future;
          }
          if (request.url.queryParameters.containsKey('cursor')) {
            paging = true;
            return older.future;
          }
          return _reply(_page([_trade(1, status: 'pending')], cursor: 'older'));
        }),
      );
      await tester.pump(const Duration(seconds: 10));
      await _pump(tester);
      expect(reconciling, isTrue);
      await tester.tap(find.text('More trades'));
      await _pump(tester);
      expect(paging, isTrue);
      reconciliation.complete(
        _reply({
          'order': {'id': _id(1), 'status': 'confirmed'},
        }),
      );
      await _pump(tester);
      older.complete(_reply(_page([_trade(2, buy: false)])));
      await _pump(tester);
      expect(find.text('Loading…'), findsNothing);
      expect(find.byKey(ValueKey('trade-${_id(2)}')), findsOneWidget);
      await _close(tester, account);
    },
  );

  testWidgets(
    'replacing the account controller clears old rows before the next response',
    (tester) async {
      final previous = _Account(),
          next = _Account()..accountId = 'next-account';
      await _mount(
        tester,
        previous,
        MockClient((_) async => _reply(_page([_trade(1)]))),
      );
      final reply = Completer<http.Response>();
      await _mount(tester, next, MockClient((_) => reply.future));
      expect(find.byKey(ValueKey('trade-${_id(1)}')), findsNothing);
      reply.complete(_reply(_page([_trade(2, buy: false)])));
      await _pump(tester);
      expect(find.byKey(ValueKey('trade-${_id(2)}')), findsOneWidget);
      await _close(tester, next);
      previous.dispose();
    },
  );

  testWidgets('returning from the stock route refreshes newly placed trades', (
    tester,
  ) async {
    final account = _Account(), returned = Completer<void>();
    var calls = 0;
    final client = MockClient((_) async {
      calls++;
      return _reply(_page([_trade(1), if (returned.isCompleted) _trade(2)]));
    });
    await tester.pumpWidget(
      MaterialApp(
        theme: productTheme(),
        home: LiveTradeHistoryScreen(
          account: account,
          origin: _origin,
          httpClient: client,
          onBack: () {},
          onOpenAsset: (assetId, mint) async {
            expect(assetId, 'nvidia');
            expect(mint, fixtures.nvidiaMint);
            await returned.future;
          },
        ),
      ),
    );
    await _pump(tester);
    await tester.tap(find.byKey(ValueKey('trade-${_id(1)}')));
    await tester.pump();
    await tester.tap(find.text('Open stock'));
    await _pump(tester);
    expect(calls, 1);
    returned.complete();
    await _pump(tester);
    expect(calls, 2);
    expect(find.byKey(ValueKey('trade-${_id(2)}')), findsOneWidget);
    await _close(tester, account);
  });
}

File: apps/api/test/live-trade-history.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import {buildApp} from '../src/app.js';
import {STOCK_TRADING_ASSETS} from '../src/stock-trading-catalog.js';
import {JUPITER_QUOTE_ASSETS} from '../src/jupiter-quote-reader.js';
import type {LiveTradeHistoryAdapters, LiveHistoryPosition} from '../src/live-trade-history.js';

const user='aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',other='bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb';
const wallet='FbP8bwmje245N5k3GTrx7BcKcDwbJNbfvEaEbF8eewY1';
function row(id=3,overrides:Record<string,unknown>={}) {
  return {id:`00000000-0000-0000-0000-${String(id).padStart(12,'0')}`,wallet,status:'confirmed',
    signature:'2'.repeat(88),createdAt:'2026-09-25T12:00:00.123456Z',updatedAt:'2026-09-25T12:00:01.654321Z',
    terms:{side:'buy',inputMint:JUPITER_QUOTE_ASSETS.USDC.mint,outputMint:STOCK_TRADING_ASSETS[0]!.mint,
      inputAmountRaw:'1000000',quotedOutputAmountRaw:'512345',minimumOutputAmountRaw:'510000'},...overrides};
}
function setup(rows:unknown[]=[]) {
  const calls:{userId:string;limit:number;before?:LiveHistoryPosition}[]=[];
  const adapters:LiveTradeHistoryAdapters={authenticate:async request=>request.headers.authorization==='Bearer signed-in'?{userId:user}:null,
    repository:{list:async(userId,limit,before)=>{calls.push({userId,limit,...(before?{before}:{})});return rows;}}};
  return {calls,adapters,app:buildApp({logger:false,liveTradeHistory:adapters})};
}
const auth={authorization:'Bearer signed-in'};

test('history requires an account, and remains available without a live execution adapter',async()=>{
  const {app,calls}=setup([row()]);
  try {
    assert.equal((await app.inject('/v1/trading/capabilities')).json().enabled,false);
    const anonymous=await app.inject('/v1/trading/history');
    assert.equal(anonymous.statusCode,401);assert.deepEqual(anonymous.json(),{code:'ACCOUNT_REQUIRED'});assert.equal(calls.length,0);
    const response=await app.inject({url:'/v1/trading/history',headers:auth});
    assert.equal(response.statusCode,200);assert.equal(response.headers['cache-control'],'no-store');
    assert.equal(response.json().orders[0].id,row().id);assert.deepEqual(calls,[{userId:user,limit:20}]);
  }finally{await app.close();}
});

test('history projects trusted asset identity and reviewed raw quantities, never wires or execution claims',async()=>{
  const data=row(3,{review:{unsignedTransaction:'private wire'},unsignedTransaction:'private wire',reviewDigest:'private digest'});
  const {app}=setup([data]);
  try {
    const result=(await app.inject({url:'/v1/trading/history',headers:auth})).json();
    assert.deepEqual(result,{schemaVersion:1,network:'solana:mainnet-beta',orders:[{
      id:data.id,wallet:data.wallet,status:'confirmed',signature:data.signature,createdAt:data.createdAt,updatedAt:data.updatedAt,
      asset:{assetId:'apple',mint:STOCK_TRADING_ASSETS[0]!.mint,symbol:'AAPLx',name:'Apple xStock',decimals:8},
      terms:data.terms,amountUnits:'raw_token_units',amountsStatus:'reviewed_quote',
    }],nextCursor:null});
  }finally{await app.close();}
});

test('history supports every submitted status and sell-side asset identity',async()=>{
  const rows=['pending','confirmed','failed','expired'].map((status,index)=>row(4-index,{status,
    terms:{...row().terms,side:'sell',inputMint:STOCK_TRADING_ASSETS[9]!.mint,outputMint:JUPITER_QUOTE_ASSETS.USDC.mint}}));
  const {app}=setup(rows);
  try {
    const response=await app.inject({url:'/v1/trading/history',headers:auth});
    assert.equal(response.statusCode,200);
    assert.deepEqual(response.json().orders.map((order:any)=>[order.status,order.asset.symbol,order.terms.side]),
      rows.map(order=>[order.status,'NFLXx','sell']));
  }finally{await app.close();}
});

test('keyset cursor preserves tied timestamps to microseconds, binds the account, and changes page size safely',async()=>{
  const rows=[row(3),row(2),row(1)];const {app,calls}=setup(rows);
  try {
    const first=await app.inject({url:'/v1/trading/history?limit=2',headers:auth});
    assert.equal(first.statusCode,200);const page=first.json();
    assert.deepEqual(page.orders.map((order:any)=>order.id),[row(3).id,row(2).id]);
    const cursor=page.nextCursor;assert.ok(typeof cursor==='string');
    rows.splice(0,2);
    const second=await app.inject({url:`/v1/trading/history?limit=1&cursor=${cursor}`,headers:auth});
    assert.equal(second.statusCode,200);assert.equal(second.json().orders[0].id,row(1).id);assert.equal(second.json().nextCursor,null);
    assert.deepEqual(calls[1],{userId:user,limit:1,before:{id:row(2).id,createdAt:row(2).createdAt}});
    const foreign=Buffer.from(JSON.stringify([1,other,row(2).createdAt,row(2).id])).toString('base64url');
    assert.equal((await app.inject({url:`/v1/trading/history?cursor=${foreign}`,headers:auth})).statusCode,400);
    assert.equal(calls.length,2);
  }finally{await app.close();}
});

test('invalid pagination is bounded before storage and cannot select another wallet/account',async()=>{
  const {app,calls}=setup();
  try {
    for(const query of ['limit=0','limit=51','limit=100000000000','limit=-1','limit=1.5','limit=01','limit=','limit=1&limit=2',
      'userId='+other,'wallet='+wallet,'cursor=','cursor=!!!','cursor='+('a'.repeat(513)),
      'cursor='+Buffer.from(JSON.stringify([1,user,'2026-02-30T12:00:00.000001Z',row().id])).toString('base64url')]){
      const response=await app.inject({url:'/v1/trading/history?'+query,headers:auth});
      assert.equal(response.statusCode,400,query);assert.deepEqual(response.json(),{code:'INVALID_REQUEST'});
    }
    assert.equal(calls.length,0);
  }finally{await app.close();}
});

test('empty history is valid; unknown adapter and storage failures are explicit and sanitized',async()=>{
  const {app,adapters}=setup();const disabled=buildApp({logger:false});
  try {
    assert.deepEqual((await app.inject({url:'/v1/trading/history',headers:auth})).json(),
      {schemaVersion:1,network:'solana:mainnet-beta',orders:[],nextCursor:null});
    assert.equal((await disabled.inject({url:'/v1/trading/history',headers:auth})).statusCode,503);
    adapters.repository.list=async()=>{throw new Error('private database credentials');};
    const response=await app.inject({url:'/v1/trading/history',headers:auth});
    assert.equal(response.statusCode,503);assert.deepEqual(response.json(),{code:'HISTORY_UNAVAILABLE'});
    adapters.repository.list=async()=>{throw new Error('ACCOUNT_REQUIRED');};
    assert.equal((await app.inject({url:'/v1/trading/history',headers:auth})).statusCode,401);
  }finally{await app.close();await disabled.close();}
});

test('malformed, unsubmitted, out-of-order or unsupported stored data never becomes financial history',async()=>{
  const rows:unknown[]=[];const {app}=setup(rows);
  try {
    for(const invalidRows of [[row(1),row(2)],[row(1),row(1)],[row(1,{status:'reviewed'})],[row(1,{signature:null})],
      [row(1,{createdAt:'2026-09-25T12:00:00.123Z'})],[row(1,{status:['pending']})],
      [row(1,{terms:{...row().terms,outputMint:wallet}})],
      [row(1,{terms:{...row().terms,inputAmountRaw:'1.5'}})],
      [row(1,{terms:{...row().terms,inputAmountRaw:'18446744073709551616'}})],
      [row(1,{terms:{...row().terms,minimumOutputAmountRaw:'999999999'}})]]){
      rows.splice(0,rows.length,...invalidRows);
      const response=await app.inject({url:'/v1/trading/history',headers:auth});
      assert.equal(response.statusCode,503);assert.deepEqual(response.json(),{code:'HISTORY_UNAVAILABLE'});
    }
  }finally{await app.close();}
});

test('configured web clients can read history through a bounded GET preflight',async()=>{
  const origin='https://trimmy.example';const {adapters,app:unused}=setup();await unused.close();
  const app=buildApp({logger:false,browserOrigins:[origin],liveTradeHistory:adapters});
  try {
    const headers={origin,'access-control-request-method':'GET','access-control-request-headers':'authorization'};
    const response=await app.inject({method:'OPTIONS',url:'/v1/trading/history?limit=20',headers});
    assert.equal(response.statusCode,204);assert.equal(response.headers['access-control-allow-methods'],'GET');
    assert.equal((await app.inject({method:'OPTIONS',url:'/v1/trading/history',headers:{...headers,'access-control-request-method':'POST'}})).statusCode,403);
    assert.equal((await app.inject({method:'OPTIONS',url:'/v1/trading/history',headers:{...headers,'access-control-request-headers':'x-trimmy-guest'}})).statusCode,403);
    assert.equal((await app.inject({method:'OPTIONS',url:'/v1/trading/order/'+row().id,headers})).statusCode,204);
    for(const url of ['/v1/trading/order/not-a-uuid','/v1/trading/order/:id','/v1/trading/order/'+row().id+'?extra=1','/v1/trading/execute']){
      assert.equal((await app.inject({method:'OPTIONS',url,headers})).statusCode,403);
    }
  }finally{await app.close();}
});
